File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
from itertools import combinations, permutations


resultSet = set()


def isSameId(c_id, b_id):
    for i in range(len(c_id)):
        if c_id[i] != b_id[i] and b_id[i] != '*':
            return False
    return True

    
def compareIds(choice_id, banned_id):

    for i in range(len(choice_id)):
        if len(choice_id[i]) != len(banned_id[i]):
            return;
        if not isSameId(choice_id[i], banned_id[i]):
            return;
    resultSet.add(choice_id)
    return;


def solution(user_id, banned_id):
    answer = 0
    banned_set = set()
    
    for b_id in permutations(banned_id):
        banned_set.add(b_id)
    
    for choice_ids in combinations(user_id, len(banned_id)):
        for b_id in banned_set:
            compareIds(choice_ids, b_id)
    answer = len(resultSet)
    return answer

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;

class Solution {
    int n, r;
    int answer = 0;
    boolean[] visited;
    boolean flag = false;
    int[] nums;
    String[] combi;
    
    public void permutations(int depth, String[] banned_id) {
        if (depth >= r) {
            
            int count = 0;
            
            for (int i = 0; i < r; i++) {
                int idx = nums[i];
                
                if (combi[idx].length() != banned_id[i].length()) continue;
                
                int count2 = 0;
                
                for (int j = 0; j < combi[idx].length(); j++) {
                    if (combi[idx].charAt(j) == banned_id[i].charAt(j)) {
                        count2++;
                    } else if (banned_id[i].charAt(j) == '*') {
                        count2++;
                    }
                }
                
                if (count2 == combi[idx].length()) {
                    count++;
                }
            }
            
            if (count == r) {
                flag = true;
            }
            
            return;
        }
        
        for (int i = 0; i < r; i++) {
            if (visited[i]) continue;
            
            visited[i] = true;
            nums[depth] = i;
            permutations(depth+1, banned_id);
            visited[i] = false;
            
            if (flag) {
                return;
            }
            
        }
    }
    
    public void combinations(int depth, int idx, String[] user_id, String[] banned_id) {
        if (depth >= r) {
            permutations(0, banned_id);
            
            if (flag) {
                answer++;
            }
            
            flag = false;
            return;
        }
        
        for (int i = idx; i < n; i++) {
            combi[depth] = user_id[i];
            combinations(depth+1, i+1, user_id, banned_id);
        }
    }
    
    public int solution(String[] user_id, String[] banned_id) {
        n = user_id.length;
        r = banned_id.length;
        
        combi = new String[r];
        visited = new boolean[r];
        nums = new int[r];
        
        combinations(0, 0, user_id, banned_id);
        
        return answer;
    }
}
